### PR TITLE
[Core] Sglang Layerwise Integration

### PR DESCRIPTION
**Pull request by Oasis-Git**
This integration provides better solution for sglang integration. Instead of non layerwise solution, the integration overlaps the computation with lmcache kv load/store operation. As a non evictable solution, the integration now has achieved similar performance with latest sgl-hiradix cache design.

If you hope to run the benchmark, please refer to `https://github.com/sgl-project/sglang/tree/main/benchmark/hicache`.

**PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to LMCache! Before submitting the pull request, please ensure the PR meets the following criteria. This helps us maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Please try to classify PRs for easy understanding of the type of changes. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Core]</code> for changes in the core LMCache logic (e.g., <code>LMCacheEngine</code>, <code>Backend</code> etc.)</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li> Please include sufficient unit tests to ensure the change is stay correct and robust. The unit and integration tests will always run and our comprehensive test will be triggered after the "full" label is tagged onto a PR.</li>
</ul>

<h3>What to Expect for the Reviews</h3>

We aim to address all PRs in a timely manner. If no one reviews your PR within 5 days, please @-mention one of KuntaiDu, ApostaC or YaoJiayi.

</details>

### CODE REVIEW

**Comment by Oasis-Git** [comment]
Here is the latest benchmark result comparison with SGLang's official hicache benchmark: https://github.com/sgl-project/sglang/tree/main/benchmark/hicache

image

File: lmcache/integration/sglang/sglang_adapter.py [hunk 1]
```diff
@@ -170,3 +197,116 @@ def reset(self):
...
+        retrieve_token_num = next(layerwise_retriever)
+        next(layerwise_retriever)
+        self.current_layer = 1
+        if retrieve_token_num is None:
+            return 0
+
+        retrieved_token_num = retrieve_token_num.item()
+        self.layerwise_retrievers.append(layerwise_retriever)
+
+        return retrieved_token_num
```
[the remark this answers is not in the document]
Comment (Oasis-Git) [reply lmcache/integration/sglang/sglang_adapter.py]: The logic is not exactly the same thus we hope to skip this advise.

File: lmcache/integration/sglang/sglang_adapter.py [hunk 2]
```diff
@@ -170,3 +196,119 @@ def reset(self):
...
+        load_mask = torch.ones_like(token_ids, dtype=torch.bool)
+        load_mask[:offset] = False
+
+        layerwise_retriever = self.lmcache_engine.retrieve_layer(
+            token_ids,
+            mask=load_mask,
+            kvcaches=self.kvcaches,
+            slot_mapping=slot_mapping,
+            sync=False,
+        )
```
[the remark this answers is not in the document]
Comment (Oasis-Git) [reply lmcache/integration/sglang/sglang_adapter.py]: Mentioned before.

File: examples/sgl_integration/README.md [hunk 3]
```diff
@@ -17,7 +17,7 @@ To start SGLang server with LMCache, run
 ```bash
 export LMCACHE_USE_EXPERIMENTAL=True
 export LMCACHE_CONFIG_FILE=lmcache_config.yaml
-python -m sglang.launch_server --model-path Qwen/Qwen2.5-14B-Instruct --port 30000 --tp 2 --page-size 32 --enable-lmcache-connector
+python -m sglang.launch_server --model-path Qwen/Qwen2.5-14B-Instruct --port 30000 --tp 2 --page-size 32 --enable-lmcache
 ```
-If you hope to run the benchmark, please refer to `https://github.com/Oasis-Git/sglang/tree/lmcache/benchmark/benchmark_lmcache`
+If you hope to run the benchmark, please refer to `https://github.com/sgl-project/sglang/tree/main/benchmark/hicache`
```
Comment (YaoJiayi) [inline examples/sgl_integration/README.md:22]: We can paste this into the PR description

File: lmcache/integration/sglang/sglang_adapter.py [hunk 4]
```diff
@@ -170,3 +199,117 @@ def reset(self):
...
+        if retrieve_token_num is None:
+            return 0
+
+        retrieved_token_num = retrieve_token_num.item()
+        self.layerwise_retrievers.append(layerwise_retriever)
+        self.layer_load_layer.append(1)
+
+        return retrieved_token_num
+
+    def load_kv(self, load_metadata: LoadMetadata) -> int:
```
Comment (YaoJiayi) [inline lmcache/integration/sglang/sglang_adapter.py]: please delete this

File: lmcache/v1/cache_engine.py [hunk 5]
```diff
@@ -591,7 +596,10 @@ def retrieve_layer(
 
                 assert None not in tasks
 
-                yield None
+                if layer_id == 0:
+                    yield torch.sum(ret_mask)
```
Comment (YaoJiayi) [inline lmcache/v1/cache_engine.py:606]: Use retrieve here

File: lmcache/integration/sglang/sglang_adapter.py [hunk 6]
```diff
@@ -170,3 +199,117 @@ def reset(self):
...
+            kvcaches=self.kvcaches,
+            slot_mapping=slot_mapping,
+            offset=store_metadata.offset,
+            sync=False,
+        )
+        next(layerwise_storer)
+        for _ in range(self.sgl_config.num_hidden_layers):
+            next(layerwise_storer)
+
+        self.lmcache_engine.lookup_unpin([lookup_id])
```
Comment (YaoJiayi) [inline lmcache/integration/sglang/sglang_adapter.py:294]: after retrieve

**Review by YaoJiayi — COMMENTED** [review]
LGTM! Just a few comments

File: examples/sgl_integration/README.md (hunk 3 above)
Comment (Oasis-Git) [reply examples/sgl_integration/README.md:22]: Done.

File: lmcache/integration/sglang/sglang_adapter.py (hunk 4 above)
Comment (Oasis-Git) [reply lmcache/integration/sglang/sglang_adapter.py]: Done in new commit

File: lmcache/v1/cache_engine.py (hunk 5 above)
Comment (Oasis-Git) [reply lmcache/v1/cache_engine.py:606]: Lookup function is retrieve will introduce performance drop. Will fix the problem in future(Add it in sgl side instead of lmcache side).

**Review by YaoJiayi — APPROVED** [review]
lgtm!

File: lmcache/v1/cache_engine.py (hunk 5 above)
Comment (YaoJiayi) [reply lmcache/v1/cache_engine.py:606]: Can we add a comment here?

**Review by YaoJiayi — APPROVED** [review]
lgtm! Could we add a small comment before merging it?
